File: ios/extra_DetailView/NASARedux/Modules/Application/Middleware/AsyncMiddleware.swift
//
//  AsyncMiddleware.swift
//  NASARedux
//
//  Created by John McIntosh on 3/1/19.
//  Copyright © 2019 Demo. All rights reserved.
//

import Foundation
import ReSwift

protocol AsyncActionable: Action {
    func postReduce(dispatch: @escaping DispatchFunction, state: AppState, getState: @escaping () -> AppState?)
}

let asyncMiddleware: Middleware<AppState> = { dispatch, getState in
    return { next in
        return { action in
            next(action)
            
            if let action = action as? AsyncActionable, let state = getState() {
                action.postReduce(dispatch: dispatch, state: state, getState: getState)
            }
        }
    }
}

File: android/app/src/_05/java/com/elliott/sxswreduxworkshopandroid/redux/middleware/AsyncMiddleware.kt
package com.elliott.sxswreduxworkshopandroid.redux.middleware

import com.elliott.sxswreduxworkshopandroid.network.NasaImageApi
import com.elliott.sxswreduxworkshopandroid.redux.AppState
import org.rekotlin.Action
import org.rekotlin.DispatchFunction
import org.rekotlin.Middleware

class AsyncMiddleware(private val imageApi: NasaImageApi) {

    internal val middleware: Middleware<AppState> = { dispatch, state ->
        { next ->
            { action ->
                val nextAction = next(action)

                when (action) {
                    is AsyncActionable -> {
                        action.postReduce(dispatch, state.invoke(), imageApi)
                    }
                }
                nextAction
            }

        }

    }
}

interface AsyncActionable : Action {
    fun postReduce(dispatchFunction: DispatchFunction, state: AppState?, imageApi: NasaImageApi)
}